plugins/autoip.rb

```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
+      end
+      auth_token = JSON.parse(response.body)["data"]["token"]
+   
+      # Check for an existing host in IPAM
+
+      puts "Checking for existing IP for host #{options[:hostname]}"
+      uri = options[:add_uri]
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("addresses/search_hostname/#{options[:hostname]}/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code == "404"
+        puts "Requesting IP for #{options[:hostname]}"
+      elsif response.code == "200"
+        abort "#{options[:hostname]} is already assigned #{JSON.parse(response.body)["data"][0]["ip"]}"
+      else
+        abort "There was an error requesting your IP address, IPAM returned code: #{response.code}, message: #{response.body}"
+      end
+
+      # Get the id of the VLAN to request address
 
-      # Get an IP from our IPAM system
-      uri = options[:add_uri].gsub(/SUBNET|HOSTNAME|USER|APIAPP|APITOKEN/, {
-                                     'SUBNET'   => options[:subnet],
-                                     'HOSTNAME' => options[:hostname],
-                                     'USER'     => username,
-                                     'APIAPP'   => options[:apiapp],
-                                     'APITOKEN' => options[:apitoken],
-                                      } )
+      uri = options[:add_uri]
       uri = URI.escape(uri)
-      uri = URI.parse(uri)
+      uri = URI.parse(uri.concat("subnets/cidr/#{options[:subnet]}"))
       http = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl = true
       request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code =="404"
+        abort "The subnet you requested #{options[:subnet]} can't be found message: #{response.body}"
+      elsif response.code != "200"
+        abort "There was an error while searching for the id of the requested subnet #{response.code}, message: #{response.body}"
+      end
+      subnetId = JSON.parse(response.body)["data"][0]["id"]
+
+      # Get the first free address in requested subnet
+
+      uri = options[:add_uri]
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("subnets/#{subnetId}/first_free/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
       response = http.request(request)
       if response.code != "200"
-        abort "There was an error requesting your IP address, IPAM returned #{response.code}"
+        abort "There was an error while requesting firs free address code: #{response.code}, message: #{response.body}"
```

Comment:
And maybe add a comma after "first free address", to make the message a little more legible?